contracts/cw-ics721-bridge/src/helpers.rs

```diff
@@ -0,0 +1,120 @@
+use crate::state::CLASS_ID_TO_NFT_CONTRACT;
+use crate::ContractError;
+use cosmwasm_std::{to_binary, Addr, Deps, DepsMut, Empty, StdResult, SubMsg, WasmMsg};
+use cw721::{NftInfoResponse, OwnerOfResponse};
+
+pub const MINT_SUB_MSG_REPLY_ID: u64 = 0;
+pub const TRANSFER_SUB_MSG_REPLY_ID: u64 = 1;
+pub const BURN_SUB_MSG_REPLY_ID: u64 = 2;
+pub const INSTANTIATE_AND_MINT_CW721_REPLY_ID: u64 = 3;
+pub const INSTANTIATE_CW721_REPLY_ID: u64 = 4;
+pub const INSTANTIATE_ESCROW_REPLY_ID: u64 = 5;
+pub const FAILURE_RESPONSE_FAILURE_REPLY_ID: u64 = 6;
+pub const BATCH_TRANSFER_FROM_CHANNEL_REPLY_ID: u64 = 7;
+pub const BURN_ESCROW_TOKENS_REPLY_ID: u64 = 8;
+
+pub fn mint(
+    deps: DepsMut,
+    class_id: String,
+    token_id: String,
+    token_uri: String,
+    receiver: String,
+) -> Result<SubMsg, ContractError> {
+    if !CLASS_ID_TO_NFT_CONTRACT.has(deps.storage, class_id.clone()) {
+        return Err(ContractError::UnrecognisedClassId {});
+    }
+
+    let class_uri = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id)?;
+    let mint_msg = cw721_base::ExecuteMsg::<Empty>::Mint(cw721_base::MintMsg::<Empty> {
+        token_id,
+        owner: receiver,
+        token_uri: Some(token_uri),
+        extension: Empty {},
+    });
+    let msg = WasmMsg::Execute {
+        contract_addr: class_uri.to_string(),
+        msg: to_binary(&mint_msg)?,
+        funds: vec![],
+    };
+    let msg = SubMsg::reply_always(msg, MINT_SUB_MSG_REPLY_ID);
+
+    Ok(msg)
+}
+
+pub fn transfer(
+    deps: Deps,
+    class_id: String,
+    token_id: String,
+    receiver: String,
+) -> Result<SubMsg, ContractError> {
+    if !CLASS_ID_TO_NFT_CONTRACT.has(deps.storage, class_id.clone()) {
+        return Err(ContractError::UnrecognisedClassId {});
+    }
+    // Validate receiver
+    deps.api.addr_validate(&receiver)?;
+
+    // No need to perform other checks as we can piggyback on cw721-base erroring for us
+
+    let class_uri = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id)?;
```

Comment:
the more I see this map, the more I am convinced that having two of these instead of one makes it harder to read/understand

Reply:
So do your other comments still stand? :upside_down_face:

- https://github.com/public-awesome/ics721/pull/22#discussion_r928137102
- https://github.com/public-awesome/ics721/pull/22#discussion_r928136813

If not, we can resolve them?

Reply:
Yeah I'm saying there should only be the one IndexedMap map like so
```
pub CHANNEL_CLASS_ID_CONTRACTS: IndexedMap<'a, (&'a str, &'a str), Addr, AddrIndexes<'a, T>>
```
Something like this. Or another alternative would be: 
```
pub const CHANNEL_CLASS_ID_CONTRACTS: Map<&str, &str), Empty> = Map::new("chanel_class_id_state");
```

Reply:
I have never successfully used an `IndexedMap` in CosmWasm, so bear with me:

I'm not entirely sure that using an `IndexedMap` here is more elegant than the double map approach. I find the double map approach to be readable / understandable. By contrast, here is the code to do a lookup of class info based on a nft address if we do an indexed map:

```rust
    #[test]
    fn test_indexed_map() {
        let mut deps = mock_dependencies();

        let expected = ClassInfo {
            class_id: "a/b/c.123".to_string(),
            nft_contract: Addr::unchecked("wasm"),
            class_uri: None,
        };

        class_info()
            .save(deps.as_mut().storage, "a/b/c.123".to_string(), &expected)
            .unwrap();

        let info = class_info()
            .idx
            .nft_contract
            .item(deps.as_mut().storage, Addr::unchecked("wasm"))
            .unwrap()
            .map(|(_, value)| value);

        assert_eq!(expected, info.unwrap());
    }
```

Using this implementation of the map:

```rust
#[derive(Serialize, Deserialize, Clone)]
pub struct ClassInfo {
    pub class_id: String,
    pub nft_contract: Addr,
    pub class_uri: Option<String>,
}

pub struct ClassInfoIndexes<'a> {
    pub nft_contract: UniqueIndex<'a, Addr, ClassInfo>,
}

impl<'a> IndexList<ClassInfo> for ClassInfoIndexes<'a> {
    fn get_indexes(
        &'_ self,
    ) -> Box<dyn Iterator<Item = &'_ dyn cw_storage_plus::Index<ClassInfo>> + '_> {
        let v: Vec<&dyn Index<ClassInfo>> = vec![&self.nft_contract];
        Box::new(v.into_iter())
    }
}

pub fn class_info<'a>() -> IndexedMap<'a, String, ClassInfo, ClassInfoIndexes<'a>> {
    let indexes = ClassInfoIndexes {
        nft_contract: UniqueIndex::new(
            |class_info| class_info.nft_contract.clone(),
            "CLASS_INFO__NFT_CONTRACT",
        ),
    };
    IndexedMap::new("CLASS_INFO", indexes)
}
```

IMO, this is pretty unreadable / hard to digest. @humanalgorithm what do you think? If you find this to improve the quality of the code I'm happy to push the refactor.

Reply:
@ezekiiel use the macro :) https://github.com/CosmWasm/cw-plus/tree/main/packages/storage-macro

Reply:
@mccallofthewild did not know that existed! Thanks!

Reply:
@humanalgorithm I know this is annoying, but getting even more into the implementation here I really don't think an `IndexedMap` is the way. Some reasons:

1. Getting the value by key from an `IndexedMap` sub-index is weird:
   ```rust
        let info = class_info()
            .idx
            .cw721_addr
            .item(deps.as_ref().storage, Addr::unchecked("wasm"))
            .unwrap()
            .map(|(_, value)| value)
            .unwrap();
   ```
2. We set the `class_uri` for a NFT we are receiving before we set the rest of the data for the `class_id` which means that there are times where some state for a class ID needs to be set but some should not be set. This makes storing all the data in one struct in an `IndexedMap` awkward: https://github.com/public-awesome/ics721/blob/d397c0677624c90708215d53380960e3b54ddb12/contracts/cw-ics721-bridge/src/contract.rs#L222-L225
3. We can't do `has` queries against anything other than the primary index in an `IndexedMap`. This makes checks like this not possible: https://github.com/public-awesome/ics721/blob/d397c0677624c90708215d53380960e3b54ddb12/contracts/cw-ics721-bridge/src/contract.rs#L298

Having three different maps here is annoying - I agree. I do think that is just "the way" here. They are not always in sync, and we need to use the suite of methods available on `Map`.

Reply:
@ezekiiel This way is a bit cleaner :)
```rust 
let (_, info) = class_info()
    .idx
    .cw721_addr
    .item(&deps.storage, key)?
    .ok_or(StdError::not_found("class_info"))?;
 ```
 For the indexed version you'd prob need to save a draft
 ```rust
pub struct ClassInfoDraft {
    pub class_id: Option<String>,
    pub cw721_addr: Option<Addr>,
    pub class_uri: Option<Option<String>>,
 }
 const DRAFT: Item<ClassInfoDraft> = Item::new("class_info_draft");
 ```
in the execution, update it in the reply, then require all + persist to class_info + delete draft in a final message.


UniqueIndex is just storing a map of index keys -> primary keys so you can recreate `.has` pretty easily :)
```rust
fn has_contract_class_info(store: &dyn Storage, nft_contract: Addr) -> bool {
    Map::<Addr, ()>::new("CLASS_INFO__NFT_CONTRACT").has(store, nft_contract)
}
```

Reply:
@ezekiiel 
Hey I think we might have gotten into some confusion here because I used the word "IndexedMap" instead of "Map". I think maybe people were thinking I was saying to use IndexedMap so we can do a reverse lookup, that's not what I intended.  

What we need is three maps, one of them being multi-key. Here are the descriptions as described in the ics721-return-trip diagram (https://www.figma.com/file/NdfNN0zOQiEyHM5G9d0uly/ics721-return-trip?node-id=0%3A1): 

We need to maintain three different maps (reduced to two after address hashing is enabled): 

  1. **nft_tracking_map:** There is one entry for every NFT that is currently being accounted for on-chain, both for locked local NFTs and remote NFTs that are going outbound. 

Format is ```(<channel_id>, <class_id>, <token_id>): <contract_address>```

2)**contract_address_to_class_id_map**: When we initiate a transfer from a contract, we only have the contract address, and we will use this to reverse lookup a previously saved class_id

Format is ```<contract_address>: <class_id>```  

3)**class_id_to_contract_addr_map**: Given a class_id and channel, locates the contract address

Format is ```<class_id>: <contract_address>```  